### DISCUSSION TITLE
XIAOGYANのArduino用ライブラリSpeaker/Speaker2クラス使い分けのためのコンパイルスイッチ廃止しませんか？

### DISCUSSION THREAD

**Discussion by y-ohe** [algyan/XIAOGYAN#49] [Ideas]
@matsujirushi  さん
XIAOGYANのArduino用ライブラリSpeaker/Speaker2クラス使い分けのためのコンパイルスイッチSPEAKERですが、無くして動的に切り換えながらどちらのクラスのAPIも使える様にしませんか？
せっかくのライブラリなのに、必ずコンパイルすることが前提のものとするより、.a形式でのリンクも可能なライブラリにもなるようにしておく方がいいかと考えています(あまりニーズは無いかも)。
Platformioでは、「platformio.ini」で定義してのビルドも比較的簡単に可能ですが、Arduino IDEでは、ライブラリソース側に対してのコンパイルスイッチの定義を受け渡すのが困難だと思われます(自分が知らないだけ？)。

具体的には、まだ構想中ですが、begin() の引数指定で切り替わるようにできれば良いのかなと、考えています。
一番大変なのは、それぞれで同一リソースを使用するので、切り換える際にリソース排他のために、リソース使用終了処理を確実に行うところになると思います。
ここは、SpeakerクラスとSpeaker2クラスを多重継承したクラスを準備して、その中で行うような実装をイメージしています。

結構大変になるし、諦めるしかないですかね。

**Comment by y-ohe** [comment]
`ここは、SpeakerクラスとSpeaker2クラスを多重継承したクラスを準備して、その中で行うような実装をイメージしています。`
単にそれぞれのクラスのオブジェクトを利用できれば良いので、多重継承はいらない気がしてきました。

**Comment by matsujirushi** [comment]
> .a形式でのリンクも可能なライブラリに

これは不要かなと思います。コンパイル時間はたいして短くならないし、コンパイル時の最適化ができなくなるので。

> begin() の引数指定で切り替わるようにできれば良いのかな

（ハード改造していたりした事情から）`Speaker`と`Speaker2`を分けていますが、`Speaker`に統合するのが良いと思います。

* #45

**Comment by matsujirushi** [comment]
`Speaker`と`Speaker2`を統合して、`SPEAKER`マクロを削除しました。
https://github.com/algyan/xiaogyan_arduino/commit/ca2c95758203c0651d47aaa5552923d57ec18c30
https://github.com/algyan/xiaogyan_examples/commit/cb62cb13d1214930523f118127ce34870e80c6b4
